docs/design_notes/lift.md

```diff
@@ -0,0 +1,355 @@
+  def wrapper(scope_fn, repack_fn, variable_groups, rng_groups, *args):
+    # normally we would first call into a JAX transformed function here...
+    target, rest = variable_groups
+    def trans(x):
+      if x.ndim == 2:
+        return x.T
+      return x
+    target = jax.tree_map(trans, target)
+    variable_groups = (target, rest)
+    scope = scope_fn(variable_groups, rng_groups)
+    y = fn(scope, *args)
+    out_variables = repack_fn(scope)
+    return y, out_variables
+  return lift.pack(wrapper, (target, variables), (variables,), (rngs,))
+
+x = jnp.ones((3, 2))
+y, params = init(lift_id(core_nn.dense))(random.PRNGKey(0), x, 4)
+```
+
+NOTE that most users should not need to interact with `pack` directly.
+Please open a GitHub issue when you find a use case that is not supported yet by the existing lifted transformations.
+
+### Supported transformations
+
+| Jax Transform | Supported in Linen? | Comments |
+|-|-|-|
+| vmap | ✅ |  |
+| scan | ✅ | Carry variables cannot be initialized inside the scan body. |
+| remat | ✅ |  |
+| jit | ✅ | Current implementation might cause unnecessary recompilation. |
+| pmap | ❌ |  |
+| grad | ❌ |  |
+| xmap | ❌ |  |
+| custom_vjp | ❌ |  |
+| custom_jvp | ❌ |  |
+
+References:
+- [Linen transforms documentation](https://flax.readthedocs.io/en/latest/flax.linen.html#module-flax.linen.transforms).
+- [Linen transforms source code](https://github.com/google/flax/blob/master/flax/linen/transforms.py)
+- [Core lifting source code](https://github.com/google/flax/blob/master/flax/core/lift.py)
+
+### Linen examples
+
+Going back to our original example, we can now use `nn.vmap` to simplify our implementation:
+
+```python
+class LinenVmap(nn.Module):
+  @nn.compact
+  def __call__(self, xs):
+    VmapMLP = nn.vmap(MLP, variable_axes={'params': 0}, split_rngs={'params': True}, in_axes=0)
+    return VmapMLP(name='mlp')(xs)
+
+variables = LinenVmap().init(random.PRNGKey(0), xs)
+print(jax.tree_map(jnp.shape, variables['params']))
+"""==>
+{
+    mlp: {
+        Dense_0: {
+            bias: (3, 4),
+            kernel: (3, 2, 4),
+        },
+        Dense_1: {
+            bias: (3, 1),
+            kernel: (3, 4, 1),
+        },
+    },
+}
+"""
+```
+
+Here we use `variable_axes={'params': 0}` to indicate that parameters are vectorized rather than shared and `split_rngs={'params': True}` means each set of parameters is initialized independently.
+
+We can also extend the example with some inner state by adding a `BatchNorm` layer:
+
+```python
+class StatefulMLP(nn.Module):
+  @nn.compact
+  def __call__(self, x, *, train):
+    h = nn.Dense(4, name='hidden')(x)
+    h = nn.BatchNorm(axis_name="batch")(h, use_running_average=not train)
+    h = nn.relu(h)
+    return nn.Dense(1, name='out')(h)
+
+class LinenStatefulVmap(nn.Module):
+  @nn.compact
+  def __call__(self, xs, *, train):
+    VmapMLP = nn.vmap(StatefulMLP, variable_axes={'params': 0, 'batch_stats': 0}, split_rngs={'params': True}, in_axes=0)
```

Comment:
I am guessing that `batch_stats` is the name of the collection that `nn.BatchNorm` is using?

Reply:
yes